src/pudl/dfc.py

```diff
@@ -0,0 +1,154 @@
+import os
+import uuid
+from pathlib import Path
+from typing import Dict, Iterator, List, Optional, Tuple
+
+import pandas as pd
+import prefect
+from prefect import task
+
+logger = logging.getLogger(__name__)
+
+
+class DataFrameCollection:
+    """This class can hold named pandas.DataFrame that are stored on disk or GCS.
+
+    This should be used whenever dictionaries of named pandas.DataFrames are passed
+    between prefect tasks. Due to the implicit in-memory caching of task results it
+    is important to keep the in-memory footprint of the exchanged data small.
+
+    This wrapper achieves this by maintaining references to tables that themselves
+    are stored on a persistent medium such as local disk of GCS bucket.
+
+    This is intended to be used from within prefect flows and new instances
+    can be configured by setting relevant prefect.context variables.
+    """
+
+    def __init__(
+            self,
+            storage_path: Optional[str] = None,
+            **data_frames: Dict[str, pd.DataFrame]):
+        """Initializes DataFrameCollection with the given set of DataFrames."""
+        storage_path = storage_path or prefect.context.get(
+            "data_frame_storage_path", None)
+        if not storage_path:
+            raise AssertionError(
+                "data_frame_storage_path needs to be set in prefect.context "
+                "in order to instantiate DataFrameCollection.")
+
+        self._storage_path = storage_path.rstrip("/")
+        self._instance_id = uuid.uuid1()
+        self._table_ids = {}  # type: Dict[str, uuid.UUID]
+
+        for name, data in data_frames.items():
+            if not isinstance(data, pd.DataFrame):
+                raise ValueError(f"Value of data frame {name} is not pandas.DataFrame.")
+        for name, data in data_frames.items():
+            self.store(name, data)
+
+    def _get_filename(self, name: str, table_id: uuid.UUID) -> str:
+        """Returns filename where the given dataframe shoudl be stored."""
+        return f"{self._storage_path}/{table_id.hex}/{name}"
+
+    def get(self, name: str) -> pd.DataFrame:
+        """Returns the content of the named dataframe."""
+        try:
+            return pd.read_pickle(self._get_filename(name, self._table_ids[name]))
+        except KeyError:
+            raise KeyError(f"Table {name} not found in the collection.")
+
+    def store(self, name: str, data: pd.DataFrame):
+        """Adds named dataframe to collection and stores its contents on disk."""
+        filename = self._get_filename(name, self._instance_id)
+        if not filename.startswith("gs://"):
+            # Do not make directories when dealing with remote storage.
+            # TODO(rousik): this is fairly crude solution and won't work
+            # for non gcs remote storage.
+            Path(os.path.dirname(filename)).mkdir(exist_ok=True, parents=True)
+        data.to_pickle(filename)
+        self._table_ids[name] = self._instance_id
```

Comment:
I don't think I'm understanding the relationships between `table_id`, `instance_id`, `filename` and table `name` or what a "reference" really means. Like can you have several different references which point at the same dataframe on disk? What is the use case for that? What happens if you try and save more than one dataframe that has the same name in the same collection?